### ISSUE TITLE
S3 multipart uploads seem difficult to verify

### ISSUE THREAD

**Issue by bjhartin** [laserdisc-io/fs2-aws#1108]
Hello,

The multipart upload pipe returns an `Option[ETag]`.  According to [AWS docs](https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html#large-object-checksums:~:text=the%20ETag%20for,these%20concatenated%20values), this is calculated as a 'checksum-of-checksums' of the individual parts, and finally the number of parts is appended, e.g. "...-3".

I don't see how to verify streaming uploads with the current API.  I can chunk a stream of bytes to the same chunk size as the uploader pipe, and calculate the checksums of each part, but then I lose them when I go `.through` the pipe.

I can think of a few options and am willing to create a PR:

- Add checksum verification at the part level, so `uploadPart` would fail if the MD5 returned by AWS doesn't match the chunk's
- Add ability to pass in a configuration for the AWS S3 client (I think PR #740 does this).  We could then make a final `get-object-attributes` call to get that checksum.
- Expose the pieces inside `uploadFileMultipart` at a more fine-grained level so we could do the verification after uploading parts

Maybe there are other options.  Here's some example code to show the problem:

```scala
      reader
        .chunkN(partSize.value * 1024 * 1024) // Carefully building the same size chunks as we'll use for the upload
        .flatMap { chunk =>
          // Here I have the checksum for the chunk.
          val digest = md5.digest(chunk.toArray)
          // Flatten the chunk back to bytes for the uploader
          fs2.Stream.emits(chunk.toList)
        }
        .through(
          // Here I'll get an Option[ETag], but I've lost what I need to compare it with
          uploader(s3, bigFileS3Key)
        )
```

Maybe this could be solved by fs2 alone, e.g. fanning out to do the checksum + upload somehow and fanning back in for the check.  I wasn't able to get it to work.

**Comment by semenodm** [comment]
Hello. I'm not sure that the digest you calculated in your example would be the same as 'checksum-of-checksums'. i think we should calculate digest inside uploader for each part and make it part of the result and do the validation once all parts are done. i think you suggested it in one of your options

**Comment by bjhartin** [comment]
@semenodm You're right, I was not clear enough.  AWS takes those digests, concatenates them, takes _their_ MD5, prints to hex and then appends the number of chunks, e.g. "...-3".  The problem is that I need to 'split' the stream to do the two 'folds', e.g. one which 'folds' the chunks down to S3 and one which 'folds' them down to the expected final ETag.

I'll edit my comments above.

**Comment by semenodm** [comment]
Closing it, since @BusyByte implemented the checksum validation
